### PR TITLE
Doc map with skip pointers

### PR DESCRIPTION
**Pull request by Benedikt1992**

### CODE REVIEW

File: cse/indexing/DocumentMapBuilder.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            self.__i += 1
+
+    def get(self, cid):
+        return self.__index[cid]
+
+    def listCids(self):
+        return [cid for cid in self.__index]
+
+    def __enter__(self):
+        return self.open()
```
Comment (SebastianSchmidl) [inline cse/indexing/DocumentMapBuilder.py]: I can't find the method `open()` of this class

File: cse/indexing/DocumentMapBuilder.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        self.__index.append((cid, pointer))
+        self.__content += 1
+        if self.__content >= PARTIAL_THRESHOLD:
+            self.__index.sort()
+            self.__save_partial_index(self.__index, self.__i)
+            self.__index = []
+            self.__content = 0
+            self.__i += 1
+
+    def get(self, cid):
```
Comment (SebastianSchmidl) [inline cse/indexing/DocumentMapBuilder.py]: do you really need this in the index builder?

File: cse/indexing/DocumentMapBuilder.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            os.remove(self.__document_map_dict)
+        dir = os.path.dirname(self.__document_map_index)
+        if os.path.isdir(os.path.join(dir, 'document_map_partials')):
+            shutil.rmtree(os.path.join(dir, 'document_map_partials'))
+
+    def close(self):
+        self.__index.sort()
+        self.__save_partial_index(self.__index, self.__i)
+        doc_dict = self._build_index_and_dict()
+        PackerUtil.packToFile(doc_dict, self.__document_map_dict, type=PackerUtil.PICKLE)
```
Comment (SebastianSchmidl) [inline cse/indexing/DocumentMapBuilder.py:36]: please also clear all internal state here, like you do in line 42 to 46

File: cse/indexing/DocumentMapBuilder.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            self.__index.sort()
+            self.__save_partial_index(self.__index, self.__i)
+            self.__index = []
+            self.__content = 0
+            self.__i += 1
+
+    def get(self, cid):
+        return self.__index[cid]
+
+    def listCids(self):
```
Comment (SebastianSchmidl) [inline cse/indexing/DocumentMapBuilder.py]: do you need this in the builder?

File: cse/indexing/DocumentMapBuilder.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            os.makedirs(os.path.join(dir, 'document_map_partials'))
+
+        with open(os.path.join(dir, 'document_map_partials', file_name + str(i)), 'w') as partial:
+            for tuple in partial_list:
+                partial.write(str(tuple[0]) + ',' + str(tuple[1]) + '\n')
+
+    def _build_index_and_dict(self):
+        doc_dict = ([], [])
+        dir = os.path.dirname(self.__document_map_index)
+        partials_dir = os.path.join(dir, 'document_map_partials')
```
Comment (SebastianSchmidl) [inline cse/indexing/DocumentMapBuilder.py]: `'document_map_partials'` magic name -> would better be a constant or variable

File: cse/indexing/DocumentMapBuilder.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            for file in os.listdir(partials_dir):
+                partial_files.append(open(os.path.join(partials_dir, file)))
+
+            for file_handle in partial_files:
+                current_lines.append(self._read_partial_line(file_handle))
+
+            while partial_files:
+                min_index = current_lines.index(min(current_lines))
+                skips += 1
+                if skips >= SKIP_SIZE:
```
Comment (SebastianSchmidl) [inline cse/indexing/DocumentMapBuilder.py:85]: this is the chunk size for which a pointer is stored in the in-memory dictionary, isn't it?

File: cse/indexing/FileIndexer.py [hunk 7]
```diff
@@ -178,7 +179,10 @@ def __processComment(self, index, cid, comment):
     )
 
     start = time.process_time()
-    FileIndexer("data", "comments.data", "articleMapping.data", "authorMapping.data", prep).index()
+    FileIndexer(".", "comments.csv", "articles.csv", "authors.csv", prep).indexPostingList()
```
Comment (SebastianSchmidl) [inline cse/indexing/FileIndexer.py:180]: may you also change the `indexPostingList()` method to build the new document map?

File: cse/reader/CommentReader.py [hunk 8]
```diff
@@ -78,15 +77,15 @@ def __silentParseToInt(self, data, default):
...
+        author = row[1]
         text = row[3].replace("\\n", "\n")
-        timestamp = row[4]
-        parentId = self.__silentParseToInt(row[5], None)
-        upvotes = self.__silentParseToInt(row[6], 0)
-        downvotes = self.__silentParseToInt(row[7], 0)
+        timestamp = row[5]
+        parentId = self.__silentParseToInt(row[4], None)
+        upvotes = 0 #self.__silentParseToInt(row[6], 0) # currently unused information
+        downvotes = 0 #self.__silentParseToInt(row[7], 0) # currently unused information
```
Comment (SebastianSchmidl) [inline cse/reader/CommentReader.py:89]: this is the _the guardian_ mapping, isn't it?
Should we really change our implementation based on their broken format?

File: cse/reader/CommentReader.py [hunk 9]
```diff
@@ -145,7 +144,10 @@ def __next__(self):
             raise stop
 
         iterRow = next(csv.reader([line], delimiter=self.__delimiter))
-        return self.__parseIterRow(iterRow, skipArticleMapping=False)
+        try:
+            return self.__parseIterRow(iterRow, skipArticleMapping=False)
+        except ValueError:
+            return self.__next__()
```
Comment (SebastianSchmidl) [inline cse/reader/CommentReader.py:151]: I think, this error handling can be kept in the code base

**Review by SebastianSchmidl — CHANGES_REQUESTED** [review]

**Comment by SebastianSchmidl** [comment]
please also update the readme to include the new document map dictionary filename

File: cse/indexing/DocumentMapBuilder.py (hunk 6 above)
Comment (Benedikt1992) [reply cse/indexing/DocumentMapBuilder.py:85]: right

File: cse/reader/CommentReader.py (hunk 8 above)
Comment (Benedikt1992) [reply cse/reader/CommentReader.py:89]: already a todo. until everythin works...

Comment (Benedikt1992) [reply cse/reader/CommentReader.py:89]: Added todo inline

File: cse/indexing/DocumentMapBuilder.py (hunk 3 above)
Comment (Benedikt1992) [reply cse/indexing/DocumentMapBuilder.py:36]: done

File: cse/indexing/FileIndexer.py (hunk 7 above)
Comment (Benedikt1992) [reply cse/indexing/FileIndexer.py:180]: done

File: cse/reader/CommentReader.py (hunk 9 above)
Comment (Benedikt1992) [reply cse/reader/CommentReader.py:151]: ok

**Comment by Benedikt1992** [comment]
readme updated

File: cse/reader/CommentReader.py (hunk 8 above)
Comment (SebastianSchmidl) [reply cse/reader/CommentReader.py:89]: please don't forget this!

**Review by SebastianSchmidl — APPROVED** [review]

**Pull request merged**